src/bigint.rs

```diff
@@ -467,146 +522,303 @@ impl Unsigned for BigUint {}
 
 forward_all_binop_to_val_ref_commutative!(impl Add for BigUint, add);
 
-impl<'a> Add<&'a BigUint> for BigUint {
+fn add2(a: &mut [BigDigit], b: &[BigDigit]) {
+    let mut b_iter = b.iter();
+    let mut carry = 0;
+
+    for ai in a.iter_mut() {
+        if let Some(bi) = b_iter.next() {
+            *ai = adc(*ai, *bi, &mut carry);
+        } else if carry != 0 {
+            *ai = adc(*ai, 0, &mut carry);
+        } else {
+            break;
+        }
+    }
+
+    assert!(b_iter.next() == None && carry == 0);
+}
+
+impl<'a> Add<&'a [BigDigit]> for BigUint {
     type Output = BigUint;
 
-    fn add(self, other: &BigUint) -> BigUint {
-        let mut sum = self.data;
+    fn add(mut self, other: &[BigDigit]) -> BigUint {
+        let new_len = cmp::max(self.data.len(), other.len()) + 1;
 
-        if other.data.len() > sum.len() {
-            let additional = other.data.len() - sum.len();
-            sum.reserve(additional);
-            sum.extend(repeat(ZERO_BIG_DIGIT).take(additional));
-        }
-        let other_iter = other.data.iter().cloned().chain(repeat(ZERO_BIG_DIGIT));
+        // resize isn't stable yet:
+        //self.data.resize(new_len, 0);
 
-        let mut carry = 0;
-        for (a, b) in sum.iter_mut().zip(other_iter) {
-            let d = (*a as DoubleBigDigit)
-                + (b as DoubleBigDigit)
-                + (carry as DoubleBigDigit);
-            let (hi, lo) = big_digit::from_doublebigdigit(d);
-            carry = hi;
-            *a = lo;
+        while self.data.len() < new_len {
+            self.data.push(0);
         }
```

Comment:
The advantage of `.extend(repeat(0).take(foo))` is that it calls `reserve` up front before pushing.  (The old code was redundant in that.)  You've used repeat.take elsewhere, why not here?